include/phoenix/script.hh

```diff
@@ -226,6 +227,46 @@ namespace phoenix {
 		const std::type_info* _m_type {nullptr};
 	};
 
+	/// \brief Represents an object associated with an instance in the script.
+	///
+	/// Instances allocated with init_opaque will be backed up by this class with plain memory storage
+	class opaque_instance final : public instance {
+	public:
+		PHOENIX_INTERNAL opaque_instance(symbol const& sym, std::vector<symbol*> const& members);
+		PHOENIX_INTERNAL ~opaque_instance();
+
+		PHOENIX_INTERNAL std::uint8_t* data() {
+			return _m_storage.get();
+		}
+
+		PHOENIX_INTERNAL const std::uint8_t* data() const {
+			return _m_storage.get();
+		}
+
+	private:
+		template <typename T, typename... Args>
+		PHOENIX_INTERNAL T* construct_at(size_t offset, Args&&... args);
+
+		std::unique_ptr<std::uint8_t[]> _m_storage;
+		std::unique_ptr<std::string*[]> _m_strings;
+		size_t _m_str_count {0};
+	};
+
+	/// \brief Represents object instance in the script with no defined backing to memory.
+	///
+	/// Expected to be used for DMA mods or to emulate variable-like access to engine-functions.
+	class transient_instance : public instance {
+	public:
+		transient_instance();
+		~transient_instance();
+
+		virtual void read32(void* data32, symbol const& sym, size_t index) = 0;
+		virtual void write32(const void* data32, symbol const& sym, size_t index) = 0;
+
+		virtual void write(std::string_view str, symbol const& sym, size_t index) = 0;
+		virtual const std::string& read(symbol const& sym, size_t index) = 0;
```

Comment:
These should be called `write_string` and `read_string` for clarification unless you rename the ones above to also simply `read` and `write? .